usr/lib/common/template.c

```diff
@@ -633,20 +673,128 @@ CK_RV template_copy(TEMPLATE *dest, TEMPLATE *src)
+     * to not break backward compatibility.
+     */
+    switch (type) {
+    case CKA_CLASS:
+    case CKA_KEY_TYPE:
+    case CKA_MODULUS_BITS:
+    case CKA_VALUE_BITS:
+    case CKA_CERTIFICATE_TYPE:
+    case CKA_VALUE_LEN:
+        return TRUE;
+    default:
+        return FALSE;
+    }
+}
+
+static CK_RV attribute_array_flatten(CK_ATTRIBUTE_PTR array_attr,
+                                     CK_BYTE **dest)
+{
+    CK_ULONG_32 long_len = sizeof(CK_ULONG);
+    CK_ATTRIBUTE_32 attr_32, element_32;
+    CK_ULONG_32 Val_32;
+    CK_ATTRIBUTE attr;
+    CK_BYTE *ptr = *dest;
+    CK_ATTRIBUTE_PTR attrs;
+    CK_ULONG num_attrs, i;
+    CK_RV rc;
+
+    if (!is_attribute_attr_array(array_attr->type))
+        return CKR_ATTRIBUTE_TYPE_INVALID;
+
+    attrs = (CK_ATTRIBUTE_PTR)array_attr->pValue;
+    num_attrs = array_attr->ulValueLen / sizeof(CK_ATTRIBUTE);
+
+    if (long_len == 4) {
+        attr.type = array_attr->type;
+        attr.ulValueLen = 0;
+        attr.pValue = NULL;
+        for (i = 0; i < num_attrs; i++)
+            attr.ulValueLen += attribute_get_compressed_size(&attrs[i]);
+
+        memcpy(ptr, &attr, sizeof(CK_ATTRIBUTE));
+        ptr += sizeof(CK_ATTRIBUTE);
+
+        for (i = 0; i < num_attrs; i++) {
+            if(is_attribute_attr_array(attrs[i].type)) {
+                rc = attribute_array_flatten(&attrs[i], &ptr);
+                if (rc != CKR_OK) {
+                    TRACE_ERROR("attribute_array_flatten failed\n");
+                    return rc;
+                }
+            } else {
+                memcpy(ptr, &attrs[i],
+                       sizeof(CK_ATTRIBUTE) + attrs[i].ulValueLen);
+                ptr += sizeof(CK_ATTRIBUTE) + attrs[i].ulValueLen;
+            }
+        }
+    } else {
+        attr_32.type = array_attr->type;
+        attr_32.pValue = 0x00;
+        attr_32.ulValueLen = 0;
+        for (i = 0; i < num_attrs; i++)
+            attr_32.ulValueLen += attribute_get_compressed_size(&attrs[i]);
+        memcpy(ptr, &attr_32, sizeof(CK_ATTRIBUTE_32));
+        ptr += sizeof(CK_ATTRIBUTE_32);
+        for (i = 0; i < num_attrs; i++) {
+            if(is_attribute_attr_array(attrs[i].type)) {
+                rc = attribute_array_flatten(&attrs[i], &ptr);
+                if (rc != CKR_OK) {
+                    TRACE_ERROR("attribute_array_flatten failed\n");
+                    return rc;
+                }
+            } else {
+                element_32.type = attrs[i].type;
+                element_32.pValue = 0x00;
+
+                if (flatten_ulong_attribute_as_ulong32(attrs[i].type) &&
+                    attrs[i].ulValueLen != 0) {
+                    element_32.ulValueLen = sizeof(CK_ULONG_32);
+
+                    memcpy(ptr, &attr_32, sizeof(CK_ATTRIBUTE_32));
+                    ptr += sizeof(CK_ATTRIBUTE_32);
+
+                    Val_32 = (CK_ULONG_32) *((CK_ULONG *) attrs[i].pValue);
```

Comment:
I don't think this goes through with ASAN checks (alignment of CK_ULONG vs. alignment of CK_BYTE).  I think the correct way would be to memcpy first into a properly aligned CK_ULONG, then assign to a properly aligned CK_ULONG_32, and then memcpy into ptr.